extensions/amp-pan-zoom/1.0/component-ts.tsx

```diff
@@ -0,0 +1,268 @@
+
+import {scale as cssScale, px, translate} from '#core/dom/style';
+
+import * as Preact from '#preact';
+import {useImperativeHandle, useRef} from '#preact';
+import {forwardRef} from '#preact/compat';
+import {ContainWrapper} from '#preact/component';
+import {useResizeObserver} from '#preact/hooks/useResizeObserver';
+
+import {useStyles} from './component.jss';
+import {usePanZoomState} from './hooks/usePanZoomState';
+
+const DOUBLE_TAP_TIME = 500; // Maximum time between tap starts
+
+const gestureConfig: Omit<UserGestureConfig, 'target'> = {
+  drag: {
+    filterTaps: true,
+    tapsThreshold: 10,
+  },
+};
+
+export type BentoPanZoomProps = {
+  children?: ComponentChildren;
+  controls?: boolean;
+  initialScale?: number | string;
+  initialX?: number | string;
+  initialY?: number | string;
+  maxScale?: number | string;
+};
+
+export type BentoPanZoomApi = {
+  transform(scale: number, x: number, y: number): void;
+};
+
+function getElementPosition(
+  clientX: number,
+  clientY: number,
+  element: HTMLElement
+) {
+  const elBounds = element./* REVIEW */ getBoundingClientRect();
+  return {
+    anchorX: clientX - elBounds.x,
+    anchorY: clientY - elBounds.y,
+  };
+}
+
+function classNames(...args: Array<string | false | null | 0>) {
+  return args.filter(Boolean).join(' ');
+}
+
+/**
+ * @return {PreactDef.Renderable}
+ */
+export function BentoPanZoomWithRef(
+  props: BentoPanZoomProps,
+  ref: Ref<BentoPanZoomApi>
+) {
+  const {
+    children,
+    controls = true,
+    initialScale,
+    initialX,
+    initialY,
+    maxScale,
+    ...rest
+  } = props;
+  const styles = useStyles();
+
+  const [state, actions] = usePanZoomState({
+    initialX,
+    initialY,
+    initialScale,
+    maxScale,
+  });
+  const isPannable = state.scale !== 1;
+  const canZoom = state.scale < state.maxScale;
+
+  const contentRef = useRef<HTMLDivElement>(null);
+  const containerRef = useRef<HTMLDivElement>(null);
+
+  const updateSizes = () => {
+    const contentBox = contentRef.current!./* REVIEW */ getBoundingClientRect();
+    const containerBox =
+      containerRef.current!./* REVIEW */ getBoundingClientRect();
+
+    actions.updateBounds({
+      containerSize: containerBox,
+      contentSize: contentBox,
+      contentOffset: {
+        x: contentBox.x - containerBox.x,
+        y: contentBox.y - containerBox.y,
+      },
+    });
+  };
+  useResizeObserver(contentRef, updateSizes);
+  useResizeObserver(containerRef, updateSizes);
+
+  useImperativeHandle(
+    ref,
+    () => /** @type {!BentoPanZoom.PanZoomApi} */ ({
+      transform: (scale: number, x: number, y: number) => {
+        actions.transform({
+          scale,
+          posX: x,
+          posY: y,
+        });
+      },
+    }),
+    [actions]
+  );
+
+  const panZoomStyles = {
+    transformOrigin:
+      px(-state.contentOffset.x) + ' ' + px(-state.contentOffset.y),
+    transform: translate(state.posX, state.posY) + cssScale(state.scale),
+  };
```

Comment:
This panZoomStyles might be worth memoizing in a `useMemo`

Reply:
I was considering that too.
However, the vast majority of rerenders will be during panning/zooming/scaling/pinching/etc ... which means these values will always be changing.  So this is probably a case where `useMemo` would probably not achieve any gains, and might just incur a shallowCompare cost.